File: IRMS_App/src/shared/downsample.ts
// shared/downsample.ts
// --- LTTB (Largest-Triangle-Three-Buckets) 抽樣 ---
//
// 2026-08-01 會議意見清單 #14:sessionsRepo.getData 無上限。25Hz × 10 分鐘 ≈ 15,000 列,
// 整包經 structured-clone 過 IPC,再為每一點跑一次 toLocaleTimeString(Intl 呼叫),
// 然後餵給 Chart.js 三個 15k 元素的陣列——分析視窗會肉眼可見地卡住。
//
// 為什麼用 LTTB 而不是均勻取樣:復健資料的臨床意義集中在「峰值」——這一下抬到幾度、
// 有沒有超過安全上限。均勻取樣會直接錯過落在取樣點之間的峰,把一次危險的超伸抹平成
// 看起來正常的曲線。LTTB 以三角形面積挑選每個桶內最具代表性的點,峰谷會被保留。
//
// 首尾點一律保留,時間軸範圍才不會縮水。

export interface Point {
  /** x 軸(時間戳的毫秒值) */
  x: number
  /** y 軸(要抽樣的指標值) */
  y: number
}

/**
 * @param data 已依 x 遞增排序的資料點
 * @param threshold 目標點數;<= 2 或大於等於原資料長度時原樣回傳
 */
export function lttb<T extends Point>(data: T[], threshold: number): T[] {
  const n = data.length
  if (threshold >= n || threshold <= 2) return data

  const sampled: T[] = [data[0]] // 首點固定保留
  // 扣掉首尾之後,剩下的 n-2 個點(索引 1 … n-2)平均分進 threshold-2 個桶。
  // 桶邊界必須從索引 1 起算:若整體偏移一格,最後一個桶會退化成空區間,
  // chosen 落在 n-1 而與固定保留的尾點重複(單元測試「輸出維持 x 遞增」即為此而設)。
  const bucketSize = (n - 2) / (threshold - 2)
  let a = 0 // 上一個被選中的點的索引

  for (let i = 0; i < threshold - 2; i++) {
    const rangeStart = Math.floor(i * bucketSize) + 1
    const rangeEnd = Math.min(Math.floor((i + 1) * bucketSize) + 1, n - 1)

    // 下一個桶的平均點,作為三角形的第三頂點;最後一個桶時退化為尾點本身
    const nextStart = rangeEnd
    const nextEnd = Math.min(Math.floor((i + 2) * bucketSize) + 1, n)
    let avgX = 0
    let avgY = 0
    const nextLen = Math.max(1, nextEnd - nextStart)
    for (let j = nextStart; j < nextEnd; j++) {
      avgX += data[j].x
      avgY += data[j].y
    }
    avgX /= nextLen
    avgY /= nextLen

    // 在本桶內挑出與 (前一選中點, 下一桶平均點) 圍成三角形面積最大的點
    let maxArea = -1
    let chosen = rangeStart
    const ax = data[a].x
    const ay = data[a].y
    for (let j = rangeStart; j < rangeEnd; j++) {
      const area = Math.abs((ax - avgX) * (data[j].y - ay) - (ax - data[j].x) * (avgY - ay))
      if (area > maxArea) {
        maxArea = area
        chosen = j
      }
    }
    sampled.push(data[chosen])
    a = chosen
  }

  sampled.push(data[n - 1]) // 尾點固定保留
  return sampled
}

File: IRMS_App_Tauri/src-tauri/src/downsample.rs
// downsample.rs — port of shared/downsample.ts's LTTB (Largest-Triangle-Three-Buckets)
// downsampling, used by sessions_repo::get_data's chart-mode path. Picks the most visually
// representative point per bucket by triangle area rather than uniform striding, so a single
// clinically-significant spike (a dangerous over-extension) survives downsampling instead of
// being averaged away between sample points. First and last points are always kept so the time
// axis range never shrinks.

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Point {
    pub x: f64,
    pub y: f64,
}

/// `data` must already be sorted by x ascending. Returns `data` unchanged if threshold is <= 2
/// or >= data.len() (a "downsample to 2 points" request is degenerate, not meaningful).
pub fn lttb(data: &[Point], threshold: usize) -> Vec<Point> {
    let n = data.len();
    if threshold >= n || threshold <= 2 {
        return data.to_vec();
    }

    let mut sampled = Vec::with_capacity(threshold);
    sampled.push(data[0]);

    // Buckets span indices 1..=n-2 (excluding the fixed first/last points), divided evenly
    // across threshold-2 buckets. Bucket boundaries must start counting from index 1: an
    // off-by-one here degenerates the last bucket to an empty range, landing "chosen" on n-1 and
    // colliding with the fixed last point (mirrors the TS test asserting strictly increasing x).
    let bucket_size = (n - 2) as f64 / (threshold - 2) as f64;
    let mut a = 0usize;

    for i in 0..(threshold - 2) {
        let range_start = (i as f64 * bucket_size).floor() as usize + 1;
        let range_end = (((i + 1) as f64 * bucket_size).floor() as usize + 1).min(n - 1);

        let next_start = range_end;
        let next_end = (((i + 2) as f64 * bucket_size).floor() as usize + 1).min(n);
        let next_len = (next_end.saturating_sub(next_start)).max(1);
        let mut avg_x = 0.0;
        let mut avg_y = 0.0;
        for j in next_start..next_end {
            avg_x += data[j].x;
            avg_y += data[j].y;
        }
        avg_x /= next_len as f64;
        avg_y /= next_len as f64;

        let ax = data[a].x;
        let ay = data[a].y;
        let mut max_area = -1.0;
        let mut chosen = range_start;
        for j in range_start..range_end {
            let area = ((ax - avg_x) * (data[j].y - ay) - (ax - data[j].x) * (avg_y - ay)).abs();
            if area > max_area {
                max_area = area;
                chosen = j;
            }
        }
        sampled.push(data[chosen]);
        a = chosen;
    }

    sampled.push(data[n - 1]);
    sampled
}

#[cfg(test)]
mod tests {
    use super::*;

    fn series(values: &[f64]) -> Vec<Point> {
        values
            .iter()
            .enumerate()
            .map(|(x, &y)| Point { x: x as f64, y })
            .collect()
    }

    #[test]
    fn below_threshold_returns_unchanged() {
        let d = series(&[1.0, 2.0, 3.0]);
        assert_eq!(lttb(&d, 10), d);
        assert_eq!(lttb(&d, 3), d);
    }

    #[test]
    fn threshold_at_or_below_two_does_not_sample() {
        let d = series(&[1.0, 2.0, 3.0, 4.0, 5.0]);
        assert_eq!(lttb(&d, 2), d);
    }

    #[test]
    fn output_length_equals_threshold() {
        let values: Vec<f64> = (0..5000).map(|i| (i as f64 / 50.0).sin() * 90.0).collect();
        let d = series(&values);
        assert_eq!(lttb(&d, 500).len(), 500);
    }

    #[test]
    fn first_and_last_points_always_kept() {
        let values: Vec<f64> = (0..1000).map(|i| i as f64).collect();
        let d = series(&values);
        let out = lttb(&d, 100);
        assert_eq!(out[0], d[0]);
        assert_eq!(*out.last().unwrap(), *d.last().unwrap());
    }

    #[test]
    fn output_x_strictly_increasing() {
        let values: Vec<f64> = (0..3000).map(|i| (i as f64 / 30.0).cos() * 45.0 + 45.0).collect();
        let d = series(&values);
        let out = lttb(&d, 300);
        for i in 1..out.len() {
            assert!(out[i].x > out[i - 1].x);
        }
    }

    #[test]
    fn preserves_an_isolated_spike() {
        // A flat line with one spike in the middle (a single dangerous over-extension) — this is
        // exactly the clinical signal that matters.
        let mut values = vec![10.0; 2000];
        values[977] = 165.0;
        let d = series(&values);
        let out = lttb(&d, 200);
        assert!(out.iter().any(|p| p.y == 165.0));
    }

    #[test]
    fn lttb_catches_a_spike_uniform_striding_would_miss() {
        let mut values = vec![10.0; 2000];
        values[977] = 165.0;
        let d = series(&values);
        let stride = (2000f64 / 200.0).ceil() as usize;
        let uniform: Vec<Point> = d.iter().step_by(stride).copied().collect();
        assert!(!uniform.iter().any(|p| p.y == 165.0)); // 977 % 10 != 0
        assert!(lttb(&d, 200).iter().any(|p| p.y == 165.0));
    }

    #[test]
    fn preserves_multiple_spikes() {
        let mut values = vec![5.0; 4000];
        for &i in &[301usize, 1123, 2517, 3733] {
            values[i] = 120.0 + (i % 7) as f64;
        }
        let d = series(&values);
        let out = lttb(&d, 400);
        for &i in &[301.0, 1123.0, 2517.0, 3733.0] {
            assert!(out.iter().any(|p| p.x == i));
        }
    }
}
